### ISSUE TITLE
Official support for restoring Kubernetes cluster

### ISSUE THREAD

**Issue by serathius** [kubernetes/kubernetes#118501]
### What would you like to be added?

I would like to propose for Kubernetes to provide an official support for a restore operation. As Kubernetes stores all it's data in etcd, this feature would be mostly about defining an official guided way for users to restore their etcd. 

## Background

Restoring Kubernetes cluster requires 2 things:
* Restoring etcd cluster using some previously created backup
* Invalidating cache for all Kubernetes clients

Etcd supports a cluster restore as described in https://etcd.io/docs/latest/op-guide/recovery/#restoring-a-cluster. Restoring cluster requires a previously saved backup which generation we leave up to user preferences. 

Restoring an etcd is essentially "going back in time" which causes a problem for controllers. Due to performance reasons Kubernetes controllers use Informers which works like a local cache for them. Etcd going back in time means that some changes cached in Informers are invalid. Not having a proper client cache invalidation strategy can lead to unpredictable and inconsistent behavior of controllers.

## Proposal
* Etcd restore operation should support bumping revision by an arbitrary number.
* Etcd restore operation should support marking old revisions as compacted.

As defined in [The Implicit Kubernetes-Etcd Contract](https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit?usp=sharing) etcd revision is used to establish a total order of operation. Kubernetes expects it to be non-decreasing. Problem is that Etcd restore operation breaks this assumption. If we could bump revision by an arbitrary number we could ensure that restore operation will not decrease the revision. For example for 1 week backup I can bump the revision by 1`000`000`000 assuming that etcd I run has less then 1500 QPS. Note: Kubernetes doesn't require revision to be continuous, it can include arbitrary gaps.

To ensure all kubernetes clients clear their cache I propose to mark to mark all revisions (including the bump) as compacted. This prevents etcd from responding to requests about revisions that were lost (executed after backup but before restore) and ensures that all clients will invalidate their cache. Using compaction window to mark revisions as unavailable seems like a natural mechanism for this case.

TODO:
* [ ] Implement optional `--bump-revision` flag to `etcdutl snapshot restore` operation.
* [ ] Implement optional `--mark-compacted` flag to `etcdutl snapshot restore` operation.
* [ ] Ensure that those operations work for etcd v3.4.X and v3.5.X
* [ ] Add e2e tests for Kubernetes restore operation.
* [ ] Document the procedure

This was originally proposed and discussed on SIG-Apimachinery meeting May 31st 2023. https://youtu.be/sueYXW89ZHg
SIG leads confirmed that there is no KEP needed as most of the work will be done on etcd side.

cc @ahrtr @wojtek-t @logicalhan @jpbetz @deads2k 

### Why is this needed?

Without zero official guidance each vendor/administrator had a different plan for etcd restore many of which were broken and incorrect. Having an official guided restore operation allow whole community to work together and share their experience leading to improvements in Kubernetes disaster recovery handling.

**Comment by ahrtr** [comment]
Please see my comment https://github.com/etcd-io/etcd/issues/16028#issuecomment-1581737360

**Comment by serathius** [comment]
> This was originally proposed and discussed on SIG-Apimachinery meeting May 31st 2023. https://youtu.be/sueYXW89ZHg
SIG leads confirmed that there is no KEP needed as most of the work will be done on etcd side.

**Comment by wojtek-t** [comment]
> To ensure all kubernetes clients clear their cache I propose to mark to mark all revisions (including the bump) as compacted. This prevents etcd from responding to requests about revisions that were lost (executed after backup but before restore) and ensures that all clients will invalidate their cache. Using compaction window to mark revisions as unavailable seems like a natural mechanism for this case.

You also need to ensure that all clients (including kube-apiserver watchcache) will actually break the watches. What is your plan for that? Just document that "all kube-apiservers have to be bring down for the etcd-restoration duration"?

**Comment by serathius** [comment]
> You also need to ensure that all clients (including kube-apiserver watchcache) will actually break the watches. What is your plan for that?  

Bumping etcd revision over the revision stored by clients and marking it as compacted will result in requests specifying exact revision to return compacted error including breaking all watches.

**Comment by serathius** [comment]
Etcd side work was finished. Anyone interested in finishing the Kubernetes one? cc @dusk125 @tjungblu

/help

**Comment by furon-kuina** [comment]
Hi, I'm a newbie k8s contributor. I was looking through the "help wanted" issues, and found this one.

Is this issue managable by a newbie? If so, I'd like to take it on.

/assign

**Comment by serathius** [comment]
Sorry, @furon-kuina this is not a good first issue. Please use [good first issue](https://github.com/kubernetes/kubernetes/labels/good%20first%20issue) to find easy issues.

**Comment by lyveng** [comment]
@serathius We run multiple k8s clusters provisioned using kubeadm in our private datacenter. We also have a recipe for etcd backup/restore. Documenting the backup/restore flow would be straight forward for us. Is that what is expected out of this ticket or does `kubeadm init` have to be updated to include a new argument for the etcd snapshot location which will be internally restored by `kubeadm init` before populating the etcd static pod manifest.

In either case, what would be the scope of the task `Add e2e tests for Kubernetes restore operation`? We have some experience debugging(went through the code) e2e tests that fail in our environment while executing the conformance tests. We can write new tests if required.

**Comment by tjungblu** [comment]
Sorry for the long radio silence. Starting next week I should have some more time to devote to this again.

In the spare meantime I've updated our OpenShift tests to use the rev bumping, unfortunately we've seen some resources still returned that should not be returned. For example: we've created a namespace _after_ taking the snapshot, yet after restore with bumping we're getting the namespace returned by the apiserver.

I suspect that there's some caching that's not watch based, or an invalidation that doesn't trigger correctly. There's some more info in https://issues.redhat.com/browse/OCPBUGS-17240, for those that are curious.

@Elbehery and I are going to pick this up next week and work through all the places where it could get cached and report back.  

---

As for the e2e tests, we have one for etcd already that asserts the api contract up to the watch breaking: https://github.com/etcd-io/etcd/blob/main/tests/e2e/ctl_v3_snapshot_test.go#L312 

If anyone else wants to volunteer to write an e2e for the apiserver, I'm also happy to assist. Feel free to start with some of our [OpenShift tests for recovery](https://github.com/openshift/origin/blob/6ee9dc56a612a4c886d094571832ed47efa2e831/test/extended/dr/recovery.go). I can also try to find some of our apiserver folks to help.

There are some challenges around ssh and running the commands, restarting etcd in the right way. It's also installation and OS dependent to a certain degree. It's certainly not for beginners, but it's not rocket science either.

**Comment by tjungblu** [comment]
Here's the PR: https://github.com/kubernetes/kubernetes/pull/120452
Had to change things up a bit, leveraged the static pod machinery instead of crio (for the most part).

Sadly I'm missing the most important part: the revision bumping. @serathius / @ahrtr I've noticed that we haven't cut a 3.5 release that includes that.

**Comment by ahrtr** [comment]
> Sadly I'm missing the most important part: the revision bumping. @serathius / @ahrtr I've noticed that we haven't cut a 3.5 release that includes that.

Yes, sorry for the delay of releasing 3.5.10. Already planned in https://github.com/etcd-io/etcd/issues/16733.

**Comment by serathius** [comment]
ping @tjungblu any progress? Do you need any help?
